crates/bevy_ecs/src/system/query.rs

```diff
@@ -528,11 +563,7 @@ impl<'w, 's, D: QueryData, F: QueryFilter> Query<'w, 's, D, F> {
     /// [`iter`](Self::iter) for read-only query items.
     #[inline]
     pub fn iter_mut(&mut self) -> QueryIter<'_, 's, D, F> {
-        // SAFETY: `self.world` has permission to access the required components.
-        unsafe {
-            self.state
-                .iter_unchecked_manual(self.world, self.last_run, self.this_run)
-        }
+        self.reborrow().into_iter()
```

Comment:
Sorry could you please explain again why we need to re-borrow?

`self.reborrow().into_iter()` creates a new Query with `reborrow()` and then borrows from that with `into_iter()`
`self.into_iter()` would simply create a QueryIter that borrows from the initital `&'a mut Query` so the lifetime is constrained by `a; is that the issue?

Reply:
`reborrow()` goes from `&'a mut Query<'w>` to `Query<'a>`, turning a borrowed `Query` into an owned one with a shorter lifetime.   

`into_iter()` takes `self` instead of `&mut self`, so it actually *consumes* the `Query`.  

We have an `&mut Query`, so we first `reborrow()` to get an owned `Query` with a shorter lifetime, and then use `into_iter()` to consume the owned query and create a `QueryIter` - still with the shorter lifetime.  

It's more complex when looking at one method, but across the whole type it means we can re-use the safety proofs in `reborrow()` and `as_readonly()`.  That lets us implement this method without `unsafe` code, and means the compiler would catch it if we tried to return `QueryIter<'w, ...>` instead of `QueryIter<'_, ...>`.